src/main/java/com/ibm/stocator/fs/swift/SwiftAPIClient.java

```diff
@@ -313,42 +314,71 @@ an request on an object (not a container) that has a trailing slash will lead
+
+      long lastModified = 0L;
+      if (metadata.containsHeader("Last-Modified")) {
+        lastModified = getLastModified(metadata.getFirstHeader("Last-Modified").getValue());
+      }
+
+      if (contentLength == 0) { // Check if directory
+        if (contentType.equals("application/directory")
+                || contentType.equals("application/octet-stream")) {
           isDirectory = true;
+        } else if (contentType.equals("application/octet-stream")) {
+          if (checkDirectory(path, hostName)) {
+            // If additional objects are returned then the object marks a directory
+            isDirectory = true;
+          }
+        } else {
+          return null;
         }
       }
       LOG.trace("{} is object. isDirectory: {}  lastModified: {}", path.toString(),
           isDirectory, lastModified);
       return new FileStatus(contentLength, isDirectory, 1, blockSize,
-              getLastModified(lastModified), path);
-    }
-    // We need to check if it may be a directory with no zero byte file associated
-    LOG.trace("Checking if directory without 0 byte object associated {}", objectName);
-    Collection<DirectoryOrObject> directoryFiles = cont.listDirectory(objectName + "/", '/',
-        "", 10);
-    if (directoryFiles != null) {
-      LOG.trace("{} got {} candidates", objectName + "/", directoryFiles.size());
-    }
-    if (directoryFiles != null && directoryFiles.size() != 0) {
-      // In this case there is no lastModified
-      isDirectory = true;
-      LOG.debug("Got object {}. isDirectory: {}  lastModified: {}", path, isDirectory, null);
-      return new FileStatus(0, isDirectory, 1, blockSize, 0L, path);
+              lastModified, path);
+    } catch (FileNotFoundException e) {
+      // If no metadata is found we need to check if it may
+      // be a directory with no zero byte file associated
+
+      LOG.trace("Checking if directory without 0 byte object associated {}", objectName);
+
+      if (checkDirectory(path, hostName)) {
+        // In this case there is no lastModified
+        isDirectory = true;
+        LOG.debug("Got object {}. isDirectory: {}  lastModified: {}", path, isDirectory, null);
+        return new FileStatus(0, isDirectory, 1, blockSize, 0L, path);
+      }
     }
     LOG.debug("Not found {}", path.toString());
     return null;
   }
 
+  private boolean checkDirectory(Path path, String hostName) throws IOException {
+    try {
+      // Using URI instead of Path constructor because the constructor trims trailing slashes
+      URI temp = new URI(path.toString() + Path.SEPARATOR);
+      Path directory = new Path(temp);
+      FileStatus[] listing = list(hostName, directory, false, true);
+      if (listing.length > 0) {
+        LOG.trace("{} got {} candidates", path + "/", listing.length);
+        return true;
+      }
+    } catch (URISyntaxException e) {
+      e.printStackTrace();
```

Comment:
@djalova Please don't print exceptions this way...use logger and throw them upper.